### ISSUE TITLE
Paint (Background Erase) and Scroll events

### ISSUE THREAD

**Issue by tadvi** [lxn/walk#153]
No easy way to attach Paint or Scrollbar event handlers to TableView, TreeView and other common components.

I have TableView. If it changes or gets scrolled I would like to update other part(s) of the screen. Problem is that there seems no easy way to attach event handlers on Paint (Background Erase) and on Scrollbar move events.

I do understand that most functionality is provided thru the use of Models, and use of Paint (Background Erase) is like a hammer. But sometimes there is a need to update other parts of GUI window when one of its controls gets Scrolled or somehow repainted.

I just do not feel like "monkey patching" this myself:
func (tv *TableView) WndProc(hwnd win.HWND, msg uint32, wParam, lParam uintptr) uintptr {
switch msg {
	case win.WM_ERASEBKGND:
   /// monkey patch here :-(

Any advice?

**Comment by lxn** [comment]
In one of my apps I am using parts of the following code to prevent periodic data reloading while scrolling:

```Go
package main

import (
	"fmt"
	"log"
	"strings"
	"unsafe"
)

import (
	"github.com/lxn/walk"
	. "github.com/lxn/walk/declarative"
	"github.com/lxn/win"
)

func main() {
	var mw *walk.MainWindow
	ctv := new(CustomTableView)

	if err := (MainWindow{
		AssignTo: &mw,
		Title:    "Custom Table View Example",
		MinSize:  Size{400, 200},
		Layout:   HBox{},
		Children: []Widget{
			TableView{
				AssignTo: &ctv.TableView,
				Columns: []TableViewColumn{
					{Name: "Normal"},
					{Name: "Loud"},
				},
				Model: []*Foo{
					{"this"},
					{"is"},
					{"some"},
					{"test"},
					{"data"},
					{"which"},
					{"should"},
					{"be"},
					{"enough"},
					{"to"},
					{"bring"},
					{"up"},
					{"a"},
					{"scroll"},
					{"bar"},
				},
			},
		},
	}).Create(); err != nil {
		log.Fatal(err)
	}

	if err := walk.InitWrapperWindow(ctv); err != nil {
		log.Fatal(err)
	}

	mw.Run()
}

type CustomTableView struct {
	*walk.TableView
}

func (ctv *CustomTableView) WndProc(hwnd win.HWND, msg uint32, wp, lp uintptr) uintptr {
	switch msg {
	case win.WM_ERASEBKGND:
		fmt.Println("WM_ERASEBKGND")

	case win.WM_NOTIFY:
		switch (*win.NMHDR)(unsafe.Pointer(lp)).Code {
		case win.LVN_BEGINSCROLL:
			nmlvs := (*NMLVSCROLL)(unsafe.Pointer(lp))
			fmt.Printf("LVN_BEGINSCROLL - nmlvs: %#v\n", nmlvs)

		case win.LVN_ENDSCROLL:
			nmlvs := (*NMLVSCROLL)(unsafe.Pointer(lp))
			fmt.Printf("LVN_ENDSCROLL - nmlvs: %#v\n", nmlvs)
		}
	}

	return ctv.TableView.WndProc(hwnd, msg, wp, lp)
}

type Foo struct {
	Normal string
}

func (f *Foo) Loud() string {
	return strings.ToUpper(f.Normal)
}

type NMLVSCROLL struct {
	hdr win.NMHDR
	dx  int32
	dy  int32
}
```

Hope this helps.

**Comment by tadvi** [comment]
This will help. Thank you very much.

**Comment by tadvi** [comment]
I get errors on lines:
case win.LVN_BEGINSCROLL:
and
case win.LVN_ENDSCROLL:

c:\godev\src\cc\customtableview.go:21: constant -180 overflows uint32
c:\godev\src\cc\customtableview.go:25: constant -181 overflows uint32

Looks like (*win.NMHDR)(unsafe.Pointer(lp)).Code evaluates to uint32 while 
win.LVN_BEGINSCROLL and win.LVN_ENDSCROLL are negative constants.

**Comment by tadvi** [comment]
Another related question: how to scroll TableView to make certain item visible.
Something like EnsureVisible. I think I have seen this somewhere in the walk codebase... sadly can not find it now.

**Comment by lxn** [comment]
You may have to update your copy of the `win` package for the constants.

`SetCurrentIndex` should do the trick. If you don't want to change the current index, you can call `tv.SendMessage(win.LVM_ENSUREVISIBLE, uintptr(index), 0)` instead. You may have to call the latter one twice, see implementation of `SetCurrentIndex`.

**Comment by tadvi** [comment]
Thank you. 

I was just wondering how you came to develop 'walk'? There is so much Windows specifics in API. 

Could you recommend some book or reference so I would not need to bug you with questions that are clearly Windows API programming?
I have Windows via C/C++ 5th Edition. But book talks mostly about threads, critical sections, etc. Not useful for GUI programming.

**Comment by lxn** [comment]
Well, when Go was announced publically, I thought it was a nice simple language and wanted to use it. But there was no desktop GUI support available back then. So I started to play around with some stuff and eventually it became more or less useful.

Before doing this, I had not that much experience programming directly against the Windows API. I went from VB6 to .NET Windows Forms with C# up to .NET 2.0, then D 1 + Tango + DWT, then Qt with C++, and after that I started with Go and walk. The biggest influence on walk was Qt. There is even a tool (ui2walk), still in the repository, which generates Go source code from Qt Designer ui files. But nowadays I wouldn't advocate using it.

Currently I am also experimenting with some stuff for Rust and Dart.

I'm afraid I can't recommend a book, because I haven't read any GUI programming related ones. MSDN is my primary source of information for this stuff.

**Comment by tadvi** [comment]
I did notice 'ui2walk' tool. When I saw it I had a different idea: it reminded me of Kivy. It is Python based   UI framework. They split GUI code declarations into separate files called "kv" files.
"kv" files actually look somewhat like 'walk' Go code: they are nested component creation declarations. The difference is that there is separate tool/processor that takes "kv" files and makes them into Python GUI code.
As a result UI creation code is separated from app logic and resulting app source code looks cleaner.

Because GUI component creation code in 'walk' gets rather repetitive and deep-nested once you get Splitters, context menus and so on. Maybe it would be reasonable to split GUI creation code into separate file(s) and have some tool that would generate/glue final UI code into the app?

Anyway, thanks for all the help.
